PXB-2336 Handle encrypted archive redo logs of 8.0.22

rahulmalik87 opened the pull request:
PXB-2336 Handle encrypted archive redo logs of 8.0.22

    Problem:
    --------
    From MySQL version 8.0.22, the archive redo log file has different disk format.
    There is one header block of 4096 bytes at the beginning of the file.

    Currently, it is unused and just an all-zero block. Even if archive redo file is
    encrypted, the encryption header from original redo file should be used.

    Because of extra 4096 byte all-zero header archive file, pxb cannot parse
    the archive redo file.

    Fix:
    ----
    PXB will check if first block is an all-zero block. If yes, we treat it as the
    archive redo file from 8.0.22 version and skip the remaining 4096-512 bytes and
    position the file cursor at the actual redo block.

    Also LSN calculation depends on file_size, so adjust LSN to file offset calculation
    to consider the extra header block for 8.0.22 archive redo file

rahulmalik87 commented:
jenkin results. https://pxb.cd.percona.com/view/PXB%208.0/job/percona-xtrabackup-8.0-param/192/

Test related to archive is unstable because of PXB-2357.

satya-bodapati commented on storage/innobase/xtrabackup/src/redo_log.cc:
```diff
@@ -776,6 +785,51 @@ void Archived_Redo_Log_Monitor::thread_func() {
       hdr_len -= n_read;
     }
 
+    /* for 8.0.22 onwards first block of redo archive file is all zeros */
```
```suggestion
    /* from 8.0.22 onwards, the first block of redo archive file is all zeros */
```

satya-bodapati commented on storage/innobase/xtrabackup/src/redo_log.cc:
```diff
@@ -35,6 +35,8 @@ Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA
 #include "xtrabackup.h"
 
 extern ds_ctxt_t *ds_redo;
+constexpr size_t QUEUE_BLOCK_SIZE = 4096;
```
add a comment for this variable. Maybe you should rename to "HEADER_BLOCK_SIZE"

satya-bodapati commented on storage/innobase/xtrabackup/src/redo_log.cc:
```diff
@@ -776,6 +785,51 @@ void Archived_Redo_Log_Monitor::thread_func() {
+      hdr_len = QUEUE_BLOCK_SIZE - OS_FILE_LOG_BLOCK_SIZE;
+      aligned_array_pointer<byte, UNIV_PAGE_SIZE_MAX> buf2;
+      buf2.create(QUEUE_BLOCK_SIZE - OS_FILE_LOG_BLOCK_SIZE);
+      while (hdr_len > 0 && !stopped) {
```
```suggestion
      /* Read remaining header of length (4096 - 512) bytes */
      while (hdr_len > 0 && !stopped) {
```

satya-bodapati commented on storage/innobase/xtrabackup/src/redo_log.cc:
```diff
@@ -776,6 +785,51 @@ void Archived_Redo_Log_Monitor::thread_func() {
+      }
+
+      hdr_len = OS_FILE_LOG_BLOCK_SIZE;
+      while (hdr_len > 0 && !stopped) {
```
```suggestion
      /* Read the actual data in the archive file */
      while (hdr_len > 0 && !stopped) {
```

satya-bodapati commented on storage/innobase/xtrabackup/src/redo_log.cc:
```diff
@@ -776,6 +785,51 @@ void Archived_Redo_Log_Monitor::thread_func() {
+      }
+      if (srv_redo_log_encrypt) {
+        aligned_array_pointer<byte, OS_FILE_LOG_BLOCK_SIZE> scratch_buf;
+        scratch_buf.create(OS_FILE_LOG_BLOCK_SIZE);
```
How is the scratch_buf freed?

rahulmalik87 replied:
(gdb) bt
#0  ut_allocator<unsigned char>::deallocate (this=0x7ffff0035930, ptr=0x7fffe80086b0 "", n_elements=0) at /home/rahul.malik/MySQL/src/x28/storage/innobase/include/ut0new.h:722
#1  0x00005555585aa686 in aligned_memory<unsigned char, 512ul>::free_memory (this=0x7ffff0035a40) at /home/rahul.malik/MySQL/src/x28/storage/innobase/include/ut0new.h:1215
#2  0x00005555585aa08d in aligned_memory<unsigned char, 512ul>::~aligned_memory (this=0x7ffff0035a40, __in_chrg=<optimized out>)
    at /home/rahul.malik/MySQL/src/x28/storage/innobase/include/ut0new.h:1173
#3  0x00005555585abeb8 in aligned_array_pointer<unsigned char, 512ul>::~aligned_array_pointer (this=0x7ffff0035a40, __in_chrg=<optimized out>)
    at /home/rahul.malik/MySQL/src/x28/storage/innobase/include/ut0new.h:1260
#4  0x00005555585a65e4 in Archived_Redo_Log_Monitor::thread_func (this=0x7fffffffd8b0) at /home/rahul.malik/MySQL/src/x28/storage/innobase/xtrabackup/src/redo_log.cc:822
#5  0x00005555585a54f7 in Archived_Redo_Log_Monitor::<lambda()>::operator()(void) const (__closure=0x7ffff0035ce0)
    at /home/rahul.malik/MySQL/src/x28/storage/innobase/xtrabackup/src/redo_log.cc:586
#6  0x00005555585a92bb in std::__invoke_impl<void, Archived_Redo_Log_Monitor::start()::<lambda()>&>(std::__invoke_other, Archived_Redo_Log_Monitor::<lambda()> &) (__f=...)
    at /usr/include/c++/7/bits/invoke.h:60

satya-bodapati commented on storage/innobase/xtrabackup/src/redo_log.cc:
```diff
@@ -776,6 +785,51 @@ void Archived_Redo_Log_Monitor::thread_func() {
+        fil_io_set_encryption(req_type, page_id_t(space->id, 0), space);
+        Encryption encryption(req_type.encryption_algorithm());
+        auto err = encryption.decrypt_log(req_type, buf, hdr_len, scratch_buf,
+                                          UNIV_PAGE_SIZE_MAX);
```
the scratch_buf is only OS_FILE_LOG_BLOCK_SIZE. UNIV_PAGE_SIZE_MAX is wrong here

satya-bodapati reviewed (changes requested):
Suggested commit message:

```
Problem:
--------
From MySQL version 8.0.22, the archive redo log file has different disk format.
There is one header block of 4096 bytes at the beginning of the file.

Currently, it is unused and just an all-zero block. Even if archive redo file is
encrypted, the encryption header from original redo file should be used.

Because of this extra 4096 all-zero header block in archive file, pxb cannot parse
the archive redo file.


Fix:
----
PXB will check if first block is an all-zero block. If yes, we treat it as the
archive redo file from 8.0.22 version and skip the remaining 4096-512 bytes and
position the file cursor at the actual redo block.

Also LSN calcualtion depends on file_size, so adjust LSN to file offset calculation
to consider the extra header block for 8.0.22 archive redo file

```

satya-bodapati reviewed (changes requested):
LGTM! one last request before approval

Can you please squash the commits with the final commit message?

Commit:
PXB-2336 Handle encrypted archive redo logs of 8.0.22

Problem:
--------
From MySQL version 8.0.22, the archive redo log file has different disk format.
There is one header block of 4096 bytes at the beginning of the file.

Currently, it is unused and just an all-zero block. Even if archive redo file is
encrypted, the encryption header from original redo file should be used.

Because of extra 4096 byte all-zero header archive file, pxb cannot parse
the archive redo file.

Fix:
----
PXB will check if first block is an all-zero block. If yes, we treat it as the
archive redo file from 8.0.22 version and skip the remaining 4096-512 bytes and
position the file cursor at the actual redo block.

Also LSN calculation depends on file_size, so adjust LSN to file offset calculation
to consider the extra header block for 8.0.22 archive redo file

satya-bodapati reviewed (approved):
LGTM!
